### PR TITLE
fix(x509.c): fixed -checkend return values

### LINKED ISSUE #28928: openssl x509 -checkend should not return 0 for expiring certificates
According to the documentation `openssl x509 -checkend` should exit with a non-zero exit code if the certificate is about to expire. With openssl 3.6 this is no longer true:

```
openssl x509 -enddate -checkend $(expr 180 \* 86400) -in $CERT && echo 'no expire'
notAfter=Nov 14 23:59:59 2025 GMT
Certificate will expire
no expire
```

with openssl 3.5 (Debian 13, same certificate)
```
openssl x509 -enddate -checkend $(expr 180 \* 86400) -in $CERT  && echo 'no expire'
notAfter=Nov 14 23:59:59 2025 GMT
Certificate will expire
```

It looks this was introduced in Commit dca67c0 (APPS/x509: add -multi option for outputting all certs found in input). After checking the expiration and setting the return code, a jump to `end_cert_loop` occurs where the return code is reset to 0 ([line 1134](https://github.com/openssl/openssl/blob/05bc60239e32a25a343dbac6aca9f67108cf0061/apps/x509.c#L1134))

Maybe `-checkend` shouldn't be usable together with `-multi` at all? Then this would be easy to fix by jumping to `end` instead.

### PR DESCRIPTION
**Pull request by snowdroppe**
Fixes #28928 

## Before

```bash
openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
0

% openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*10]; echo $?
Certificate will expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
Certificate will not expire
Certificate will not expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*1]; echo $?
Certificate will expire
Certificate will not expire
Certificate will not expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*4]; echo $?
Certificate will expire
Certificate will not expire
Certificate will expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*20]; echo $?
Certificate will expire
Certificate will expire
Certificate will expire
0
```

## After

```bash
openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
0

% openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*10]; echo $?
Certificate will expire
1

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
Certificate will not expire
Certificate will not expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*1]; echo $?
Certificate will expire
Certificate will not expire
Certificate will not expire
1

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*4]; echo $?
Certificate will expire
Certificate will not expire
Certificate will expire
1

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*20]; echo $?
Certificate will expire
Certificate will expire
Certificate will expire
1
```

### CODE REVIEW

**Comment by snowdroppe** [comment]
Force push to comply with styling guide `k+1 < ...` -> `k < ... - 1`

**Review by beldmit — DISMISSED** [review]
First, I'm not sure it's worth introducing this change.
To distinguish an error because of, say, wrong format, you still need to parse output.

Second, if this change is introduced, we need to document it

**Comment by snowdroppe** [comment]
> First, I'm not sure it's worth introducing this change. To distinguish an error because of, say, wrong format, you still need to parse output.

The current documentation reads `Checks if the certificate expires within the next arg seconds and exits non-zero if yes it will expire or zero if not.` which holds for versions prior to the implementation of `-multi` but is uniformly `0` after.

I agree that return codes may mask other errors and am happy to take the organisation's direction on how it should behave.

> 
> Second, if this change is introduced, we need to document it

Valid point as a lot of scripts currently depend on the return code to check for expired / near expired certs. I'll update it according to what is agreed.

Is https://github.com/openssl/openssl/issues/28928 a better thread to discuss the intent?

File: apps/x509.c [hunk 1]
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
...
-        ret = X509_cmp_time(X509_get0_notAfter(x), &tcheck) < 0;
-        if (ret)
+        if (expired)
             BIO_printf(out, "Certificate will expire\n");
         else
             BIO_printf(out, "Certificate will not expire\n");
-        goto end_cert_loop;
+
+        if (multi && k > 0)
+            ret = (ret | expired);
```
Comment (shahsb) [inline apps/x509.c]: Could be done something like following:

```
int expired = X509_cmp_time(X509_get0_notAfter(x), &tcheck) < 0;
ret = expired;  /* For single or first cert */
/* ... print ... */
if (multi && k > 0) {
    ret |= expired;  /* Accumulate */
}
```

File: apps/x509.c [hunk 2]
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
...
+        else
+            ret = expired;
+
+        if (multi && k < sk_X509_num(certs) - 1)
+            goto end_cert_loop;
```
Comment (shahsb) [inline apps/x509.c:1114]: could we please update documentation for this change?

(doc/man1/openssl-x509.pod) claims: "If the certificate will expire within the given number of seconds, then exit code 1 is returned." But it doesn't cover -multi. The PR's change aligns with this but breaks legacy scripts assuming always-0

File: apps/x509.c [hunk 3]
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
...
+        else
+            goto end;
```
Comment (shahsb) [inline apps/x509.c:1116]: Suggestion: Could you please add `-checkend` examples to the pod, mirroring the PR's CLI snippets.

File: apps/x509.c [hunk 4]
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
...
         else
             BIO_printf(out, "Certificate will not expire\n");
-        goto end_cert_loop;
+
+        if (multi && k > 0)
+            ret = (ret | expired);
+        else
+            ret = expired;
+
+        if (multi && k < sk_X509_num(certs) - 1)
```
Comment (shahsb) [inline apps/x509.c:1113]: Alternatives considered? Could add a new flag (e.g., -strict-checkend), but direct fix is simpler.

File: apps/x509.c [hunk 5]
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
...
 
-        ret = X509_cmp_time(X509_get0_notAfter(x), &tcheck) < 0;
-        if (ret)
+        if (expired)
             BIO_printf(out, "Certificate will expire\n");
         else
             BIO_printf(out, "Certificate will not expire\n");
-        goto end_cert_loop;
+
+        if (multi && k > 0)
```
Comment (shahsb) [inline apps/x509.c:1108]: Could you please add some automated tests for this change?

The OpenSSL test suite (in test/recipes/25-test_x509.t or similar) should be extended with:
+ Unit tests for single vs. multi mode.
+ Cases for exactly-at-boundary expiration (e.g., offset == exact seconds to notAfter).
+ Error handling (e.g., invalid cert input → exit 1? Currently, it might still print but exit 0—align with manpage).

**Review by shahsb — COMMENTED** [review]
Thanks @snowdroppe for your efforts on this!

It'll LGTM after the suggested fixes!

File: apps/x509.c (hunk 1 above)
Comment (snowdroppe) [reply apps/x509.c]: Agree on the shorthand for the bitwise or shorthand assignment.

I disagree on the unconditional assignment of `ret = expired;` as this will execute for all certs in the loop, overwriting previous return codes. This would result in the final return being entirely dependent on the final certificate only.

```c
if (multi && k > 0)
    // 2nd or later cert in chain
    ret |= expired;
else
    // Single cert or 1st in chain
    ret = expired;
```

File: apps/x509.c (hunk 2 above)
Comment (snowdroppe) [reply apps/x509.c:1114]: Old man page entry:

>Checks if the certificate expires within the next _arg_ seconds and exits nonzero if yes it will expire or zero if not.

Proposed new man page entry:

>Without **-multi** checks if the certificate expires within the next _arg_ seconds and exits non-zero if it will expire or zero if not.
>
>With **-multi** checks if any certificate in the input will expire within the next _arg_ seconds and exits non-zero if any will expire or zero if none will.

File: apps/x509.c (hunk 3 above)
Comment (snowdroppe) [reply apps/x509.c:1116]: Added

>Check if any certificates in a chain are due to expire within the next 30 days (returns zero if none will expire, non-zero if any will expire):
>
>`openssl x509 -in chain.pem -multi -checkend $[3600*24*30] \`
> `        && echo 'perform renewal' || echo 'renewal unnecessary'`

File: apps/x509.c (hunk 4 above)
Comment (snowdroppe) [reply apps/x509.c:1113]: I'm interested in your idea, how do you envision the behaviour under something like `-strict-checkend` vs `-checkend` ?

File: apps/x509.c (hunk 5 above)
Comment (snowdroppe) [reply apps/x509.c:1108]: Added unit tests for:

1. single + not expiring => 0
2. single + expiring => !0
3. multi + none expiring => 0
4. multi + 1st (first) expiring => !0
5. multi + 2nd (middle) expiring => !0
6. multi + 3rd (last) expiring => !0
7. multi + all expiring => !0

The edge case where offset == (notAfter - now()) is hard to test as now() is dynamic so we need a way to fix it or risk unreliable test results. Approaches I considered:

1. Hook syscall to fake the current time e.g. `faketime` which sets LD_PRELOAD
>  Adds dependencies and becomes platform dependent
2. Implement `-attime` as used for unit tests in `openssl verify`
>  Seems excessive to add functionality just to aid a single edgecase unit test for `x509`
3. Write a c unit test like in `test/X509_time_test.c:287`
>  Adds very little assurance as the existing test already checks X509_cmp_time() well
4. YOLO calculate the exact seconds in perl, pass it as an argument to `app` and hope perl now() and `app` now() doesn't span a second boundary
> Non zero chance of test failure due to being unlucky


Open to suggestions or accepting the cost isn't worth the value add.

**Comment by snowdroppe** [comment]
Pushed new commit which rebases to master and implements the above suggestions. Pending review, I will rebase again and squash into a single commit.

File: apps/x509.c (hunk 4 above)
Comment (shahsb) [reply apps/x509.c:1113]: Thanks for the ping -- happy to elaborate! The `-strict-checkend` idea was mostly a nod to minimizing breakage for folks with brittle scripts relying on the buggy "always 0" exit code (though I agree it's a corner case, and the direct fix is cleaner long-term). Here's how I'd envision the split if we went that route:

### Behavior Breakdown
- **`-checkend <seconds>` (legacy/default mode)**:
  - Prints per-cert messages as before (e.g., "Certificate will expire" or "will not").
  - Exit code: Always `0` (success), regardless of expiration. This preserves the pre-3.0 behavior and avoids script surprises.
  - In multi-mode: Still prints all messages, but exit remains `0`.
  - Use case: Interactive/one-off checks where you parse stdout for details, not exit codes.

- **`-strict-checkend <seconds>` (new strict mode)**:
  - Same printing as `-checkend`.
  - Exit code: `0` if *no* certificates expire within `<seconds>`; `1` if *any* does (bitwise OR accumulation across certs in multi-mode).
  - In multi-mode: Exit `1` if at least one cert expires; otherwise `0`. (This matches the PR's fix and the manpage intent.)
  - Use case: Automation/scripts wanting a true "expiration alert" signal (e.g., cron job: `openssl x509 -multi -in chain.pem -strict-checkend 86400 || notify-admin`).

### Why This Split?
- **Backward compat**: No behavior change for existing `-checkend` users.
- **Opt-in strictness**: New flag for those needing the "fail on expiration" semantic without rewriting scripts.
- **Docs**: Update the manpage with a clear section on both, plus examples like your CLI snippets (maybe add a multi-mode one for strict).
- **Implementation**: Minimal—gated by a new flag var (e.g., `int strict_check = 0;`), then `if (strict_check) { ret |= expired; } else { ret = 0; }` in the loop. Your current diff could adapt easily.

That said, I still lean toward the direct fix (merge as-is, deprecate the always-0 behavior in a future major), since it's a clear bug and the manpage already promises exit `1` on expiration. But if we're worried about ecosystem impact, the flag route is a low-risk hedge. Thoughts? Happy to mock up a patch if helpful.

File: apps/x509.c (hunk 2 above)
Comment (shahsb) [reply apps/x509.c:1114]: LGTM. Thanks for making the changes!

File: apps/x509.c (hunk 3 above)
Comment (shahsb) [reply apps/x509.c:1116]: Thanks @snowdroppe for making the changes!

File: apps/x509.c (hunk 1 above)
Comment (shahsb) [reply apps/x509.c]: Thanks @snowdroppe!

You're absolutely right -- my earlier unconditional ret = expired; would overwrite on every iteration, leaving only the last cert's status. Excellent catch, and your if/else structure nails it perfectly.

File: apps/x509.c (hunk 5 above)
Comment (shahsb) [reply apps/x509.c:1108]: Thanks @snowdroppe -- awesome to see the tests land so quickly! The 7 cases you added look spot-on for core coverage: they hit the happy paths, the "any expires" OR logic, and positional variations in multi-mode without overcomplicating things. If they're wired into `25-test_x509.t` (or wherever fits best), that should catch any regressions handily. Quick nit: If not already, a comment block in the test file summarizing "checkend exit codes" would make it self-documenting for future maintainers.

On the boundary edge case (offset == exact remaining seconds): Totally get the pain—`time(NULL)` is a moving target, and forcing precision there risks flakiness or bloat. Of your options:
- 1. (faketime/LD_PRELOAD): Nah, too fragile across CI platforms.
- 2. (-attime): Overkill, as you said—save that for verify if needed.
- 3. (C unit test): Fair point; `X509_time_test.c` already validates the cmp logic, so layering a Perl app test on top adds marginal value.
- 4. (YOLO Perl calc): This is my pick for minimal viable -- it's cheap, and the odds of a 1-second drift causing sporadic failures are low (especially if you run it in a tight loop or retry on mismatch). Something like:
  ```perl
  my $now = time();
  my $not_after = ...;  # From cert parse
  my $exact_offset = $not_after - $now;
  run(app([qw(-checkend), $exact_offset]), 0);  # Expect 0 (not expired)
  ```
  If it flakes once in a blue moon, we can always disable/revisit. But honestly, if the functional tests pass consistently, the cost *isn't* worth the value -- `X509_cmp_time` is battle-tested, and this is more about app plumbing.

One gap from my ask: Error handling for invalid cert input (e.g., malformed PEM → should exit 1 per manpage, not 0). Is that folded into existing tests, or worth a quick 8th case like `invalid cert + checkend → !0`? If it's already aligned (or low-risk), no sweat.

With tests in place, this is merge-ready IMO. Feel free to squash/push. Great collaboration -- thanks for iterating so nimbly! 👍

File: apps/x509.c (hunk 2 above)
Comment (ArchangeGabriel) [reply apps/x509.c:1114]: @shahsb What legacy scripts assuming always-0? The previous behaviour for years has been returning 1 on expiration. I’ve been relaying on this for more than three years (https://github.com/ArchangeGabriel/acme-user/blame/main/acme-renew.sh#L5).

File: apps/x509.c (hunk 5 above)
Comment (snowdroppe) [reply apps/x509.c:1108]: Greatly appreciate the constructive and articulate feedback!

I added a comment block preceding the unit tests which reads as follows:
> Tests for -checkend including -multi
> Discussed in https://github.com/openssl/openssl/pull/29155


I totally agree with the above sentiment and have added 2 new unit tests as recommended:

1. Naive perl-based boundary test at offset == time until expiry.
I performed 10,000 loops on a slow raspberry pi and the unlucky false positive occured 0 times. Nonetheless, I added the following comment block before the test if it ever becomes problematic in the future.
> Single + expiring at boundary
> Test may fail erroneously due to sequential now() calls
> See https://github.com/openssl/openssl/pull/29155

2. Preservation of malformed x509 PEM parsing returning non-zero.
I agree that this is a good test to have seeing as previous modifications affecting return code logic created this bug in the first place.
The test attempts `openssl x509 -in ca-key.pem -checkend 60" which should fail given the PEM contains no valid certificate. Is there a preferred malformed PEM I should use instead?

All tests run successfully locally and will rebase, squash and push as agreed.

**Comment by snowdroppe** [comment]
I believe the PR is now in a fit state for review.
I have squashed changes into a single commit and rebased to master for easier merging.

Please let me know if there are any additional actions on me. 
Also thanks to everyone for helping me contribute!

File: apps/x509.c (hunk 5 above)
Comment (shahsb) [reply apps/x509.c:1108]: Great. Thanks, @snowdroppe !!

**Comment by snowdroppe** [comment]
The unit tests trigger an overflow on the checkend `time_t` type i.e. year 2038 bug on some build platforms so I will need to revisit this. I'll update once I have a solution.

**Comment by snowdroppe** [comment]
Fixed workflow issues:
- doc-nits: non-zero -> nonzero
- unit test:
Previous -checkend overflowed time_t due to ~100 year delta.
Now generates certs in the near future (200 days) to test against.

**Comment by snowdroppe** [comment]
@beldmit Could you kindly re-approve the workflow checks please? I believe I addressed the previous erros. Thanks!

**Comment by snowdroppe** [comment]
@beldmit I believe the workflow checks are successful and this is ready to be considered for merge. Please let me know if there's anything else I should do.

The failing style check is unrelated to the changes within this PR:
`providers/implementations/kdfs/pbkdf2.c:320:indent = 12 != 8`

File: apps/x509.c (hunk 4 above)
Comment (t8m) [reply apps/x509.c:1113]: No @shahsb This is a clear regression in 3.6 version. Just needs to be fixed.

**Comment by t8m** [comment]
This needs to go into 3.6 branch as well.

**Comment by t8m** [comment]
Squashed and merged to the master and 3.6 branches. Thank you for your contribution.
